cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/ArtifactLocalizer.java

```diff
@@ -0,0 +1,249 @@
+   */
+  public File getAndUnpackArtifact(ArtifactId artifactId) throws Exception {
+    LOG.debug("Unpacking artifact {}", artifactId);
+    File jarLocation = getArtifact(artifactId);
+    File unpackDir = getUnpackLocalPath(artifactId, Long.valueOf(jarLocation.getName().split("\\.")[0]));
+    LOG.debug("Got unpack directory as {}", unpackDir);
+    if (!unpackDir.exists()) {
+      LOG.debug("Unpack directory doesnt exist, creating it now");
+      BundleJarUtil.unJar(jarLocation, unpackDir);
+    }
+    return unpackDir;
+  }
+
+  private File fetchArtifact(ArtifactId artifactId, String url,
+                             Long lastModifiedTimestamp) throws IOException, ArtifactNotFoundException {
+    HttpURLConnection urlConn = remoteClient.openConnection(HttpMethod.GET, url);
+    if (lastModifiedTimestamp != null) {
+      LOG.debug("Found existing local version with timestamp {}", lastModifiedTimestamp);
+      ZonedDateTime lastModifiedDate =
+        ZonedDateTime.ofInstant(Instant.ofEpochMilli(lastModifiedTimestamp), ZoneId.systemDefault());
+      urlConn.setRequestProperty(HttpHeaderNames.IF_MODIFIED_SINCE, lastModifiedDate.format(DATE_TIME_FORMATTER));
+    }
+
+    if (urlConn.getResponseCode() != HttpURLConnection.HTTP_OK) {
+      switch (urlConn.getResponseCode()) {
+        // If we get this response that means we already have the most up to date artifact
+        case HttpURLConnection.HTTP_NOT_MODIFIED:
+          LOG.debug("Call to app fabric returned NOT_MODIFIED");
+          urlConn.disconnect();
+          return getArtifactJarLocation(artifactId, lastModifiedTimestamp);
+        case HttpURLConnection.HTTP_NOT_FOUND:
+          throw new ArtifactNotFoundException(artifactId);
+        default:
+          String err = CharStreams.toString(new InputStreamReader(urlConn.getErrorStream(), StandardCharsets.UTF_8));
+          throw new IOException(
+            String.format("Failed to fetch artifact %s from app-fabric due to %s", artifactId, err));
+      }
+    }
+
+    Map<String, List<String>> headers = urlConn.getHeaderFields();
+    if (!headers.containsKey(LAST_MODIFIED_HEADER) ||
+      headers.get(LAST_MODIFIED_HEADER).size() != 1) {
+      //Not sure how this would happen since this endpoint should always set the header.
+      // If it does happen we should retry.
+      throw new RetryableException(String
+                                     .format("The response from %s did not contain the %s header.", url,
+                                             LAST_MODIFIED_HEADER));
+    }
+
+    ZonedDateTime newModifiedDate = LocalDateTime
+      .parse(headers.get(LAST_MODIFIED_HEADER).get(0), DATE_TIME_FORMATTER)
+      .atZone(ZoneId.systemDefault());
+    Long newTimestamp = newModifiedDate.toInstant().toEpochMilli();
+    File newLocation = getArtifactJarLocation(artifactId, newTimestamp);
+
+    try (InputStream in = urlConn.getInputStream()) {
+      FileUtils.copyInputStreamToFile(in, newLocation);
```

Comment:
This can have a race condition (one thread is copying while another thread see's the file already exists). You should write to a temp file followed by atomic rename. Or otherwise, you need a per artifact lock.

Reply:
done, it will now download to a temp file, then rename it